### ISSUE TITLE
Increase in apiserver/c-m mem usage by 100-200 MB in 100-node cluster

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#60500]
Starting sometime around yesterday night, we're observing a significant increase in apiserver memory and CPU usage. There was a sharp increase across runs 11168 and 11169 of our [100-node performance test](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce):

[from:](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11168/artifacts/ResourceUsageSummary_density_2018-02-26T17:34:31Z.json)

```
{
      "Name": "kube-apiserver-e2e-big-master/kube-apiserver",
      "Cpu": 0.909674846,
      "Mem": 830136320
},
```

[to:](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11169/artifacts/ResourceUsageSummary_density_2018-02-26T18:53:04Z.json)

```
{
      "Name": "kube-apiserver-e2e-big-master/kube-apiserver",
      "Cpu": 1.447254278,
      "Mem": 2031329280
},
```

From the [diff](https://github.com/kubernetes/kubernetes/compare/c19f8fcea...456ebf5de), I'm almost sure it's https://github.com/kubernetes/kubernetes/pull/60076 which is introducing buffering to audit-logging. I'll try to confirm this.

/assign
/cc @kubernetes/sig-scalability-bugs @wojtek-t

**Comment by shyamjvs** [comment]
FTR - Some runs which failed due to exceeding our thresholds:

- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11169
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11170
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11171
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11172
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11173

**Comment by wojtek-t** [comment]
@crassirostris @sttts

**Comment by shyamjvs** [comment]
So it indeed seems to be the audit-logging change. I looked at the memory allocation profiles from the density test [before](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11168/artifacts/profiles/ApiserverMemoryProfile_density.pdf) and [after](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11169/artifacts/profiles/ApiserverMemoryProfile_density.pdf) the change. And the allocations made by `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.audit.go: WithAudit()` went from ~6.4GB to ~23GB.

**Comment by sttts** [comment]
#60076 does only introduce the buffered backend, but does not wire it in anyway. https://github.com/kubernetes/kubernetes/pull/60237 is still not merged.

**Comment by crassirostris** [comment]
Yup, https://github.com/kubernetes/kubernetes/pull/60076 doesn't add any dependency on the new package anywhere

**Comment by shyamjvs** [comment]
I see.. I'm not sure though how that explains the increase of memory allocations from 6GB -> 24GB related to audit-logging. As I mentioned in https://github.com/kubernetes/kubernetes/issues/60500#issuecomment-368855575.

**Comment by crassirostris** [comment]
@shyamjvs I doesn't, there has to be something else. Note, that memory allocation changed 20GB -> 50GB, not 20GB -> 37GB, so the problem is not isolated in audit logging

**Comment by wojtek-t** [comment]
Looking into PR itself indeed it seems like a code that is not used anywhere.

@shyamjvs - let me take a look into that too.

**Comment by shyamjvs** [comment]
Sure.. Let me verify if it was some false positive (though at least 5 runs seemed to fail continuously in a row due to it).

**Comment by shyamjvs** [comment]
Ok.. So from perf-dash, I'm seeing that there was a transient spike in the mem-usage for few runs during that time, but it went down later.

![density_apiserver_mem_usage](https://user-images.githubusercontent.com/4333971/36730706-aceeb36a-1bc8-11e8-9a21-58f55d2f2de3.png)

**Comment by wojtek-t** [comment]
If you look into graph for load test:
- there is also a spike
- but this is shifted a little bit (it started at run 1165)
- and there is another spike that is happening now

I think this may be something real - I wouldn't close that now.

**Comment by shyamjvs** [comment]
Ok.. Let's wait to gather some more evidence.

**Comment by wojtek-t** [comment]
See this:

![screenshot from 2018-02-27 14-25-10](https://user-images.githubusercontent.com/10743879/36731218-5e09a30c-1bca-11e8-93ad-27f0ee635dfb.png)

**Comment by tpepper** [comment]
Since this is looking relatively severe, as ya'll continue to triage please add a priority and labels for milestone v1.10 to the issue and any PRs as applicable since the indicated problem commit range is in the 1.10 branch.

**Comment by shyamjvs** [comment]
Update: It seems like we saw a couple of spikes around that range, following that things seem to have gone back to normal (for about 40 runs). That said, I'd like to wait a bit more to gather more confidence. Here are apiserver mem usage graphs for density and load test resp. :

![density_apiserver_mem](https://user-images.githubusercontent.com/4333971/36840843-bc8931cc-1d46-11e8-9d73-db18d6d00f25.png)
![load_apiserver_mem](https://user-images.githubusercontent.com/4333971/36840848-c0da0472-1d46-11e8-9d44-3e3f10d99b38.png)

**Comment by shyamjvs** [comment]
Note that we're [seeing flaky failures](https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability?before=11240) for that job from time to time, with the following reason:

```
Error while waiting for replication controller density-latency-pod-69 pods to be running: Timeout while waiting for pods with labels "name=density-latency-pod-69,type=density-latency-pod" to be running
```

That 'might' be unrelated to this particular issue, but it seems like a regression. Still, it needs fixing.

**Comment by jberkus** [comment]
Adding a real priority:

/priority critical-urgent

If this is going to be in the milestone until we have more evidence, it should have status/approved-for-milestone.  So please decide one way or the other on the next test run.  Thanks!

**Comment by jberkus** [comment]
@shyamjvs given the results in #60589, can we consider this a confirmed problem for 1.10?

**Comment by shyamjvs** [comment]
Yes, I would say so. There seems to be a regression - let's approve this issue for the milestone, while I continue to investigate further into it.

**Comment by jberkus** [comment]
@shyamjvs see update on scalability test flake, which seems to be fixed.  Are we still seeing increased memory usage?

**Comment by shyamjvs** [comment]
Yes, we're still seeing flakes here (though strangely they're around increased mem for controller-manager now):

```
Container kube-controller-manager-e2e-big-master/kube-controller-manager is using 503.12109375/500 MB of memory
not to have occurred
```

(ref: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11357)

The other kinds of flakes seen are those I mentioned in https://github.com/kubernetes/kubernetes/issues/60500#issuecomment-369555508.
It strongly feels like a regression to me. I'd suggest marking this as a blocker for the release while we continue investing.

**Comment by jberkus** [comment]
@shyamjvs it's already a blocker.  Please keep us updated, thanks!

**Comment by shyamjvs** [comment]
I guess now it is? :)

**Comment by shyamjvs** [comment]
ACK.  In progress
ETA: 9/3/2018
Risks: Not yet understood well

**Comment by jberkus** [comment]
@shyamjvs update?

**Comment by shyamjvs** [comment]
@jberkus Unfortunately, I was caught up with a couple of serious regressions under https://github.com/kubernetes/kubernetes/issues/60589. Let me take a deeper look into this now.

ACK. In progress
ETA: <unknown>
Risks: Not yet understood well

**Comment by shyamjvs** [comment]
I went through the last 12 flaky failures of our [100-node performance test](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce) and they all fall under one of the 2 categories below:

1. `Error while waiting for replication controller density-latency-pod-54 pods to be running: Timeout while waiting for pods with labels "name=density-latency-pod-54,type=density-latency-pod" to be running`

E.g:
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11568
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11567
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11560
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11553
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11545
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11528
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11507
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11497
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11492

2. `Container kube-apiserver-e2e-big-master/kube-apiserver is using 1528.9296875/1500 MB of memory`

E.g:
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11546
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11529
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/11495 (in this one controller-manager also crosses the allowed threshold)

**Comment by shyamjvs** [comment]
Wrt the second kind of failures,
I used some script-ery to obtain the graph of apiserver 99%ile mem-usage over a longer time horizon (~1 month) and found this:

![apiserver_mem_usage](https://user-images.githubusercontent.com/4333971/37302877-0ecb11d2-262d-11e8-99fb-43d0bfe6d660.png)

**Comment by shyamjvs** [comment]
There was a jump which just crosses our threshold of 1500MB somewhere around run 11250 (look at the red circle). I'll try to zoom in further to narrow down this range.

**Comment by shyamjvs** [comment]
Narrowed down to a smaller range, and it seems likely the increase has been introduced b/w runs 11250 and 11255:

![narrower_range](https://user-images.githubusercontent.com/4333971/37303655-7213e80c-262f-11e8-9199-e9b902183065.png)

**Comment by shyamjvs** [comment]
And the [diff](https://github.com/kubernetes/kubernetes/compare/46f05b0c2bc153...209cdd90489a5a) has 2 commits, and I suspect the following change to be likely causing it:

https://github.com/kubernetes/kubernetes/pull/60237 Audit use buffered backend

Unlike the one we earlier suspected wrt audit-logging (https://github.com/kubernetes/kubernetes/pull/60076) that turned out to be just a red herring, the above PR actually seems to enable the buffered backend for audit-logging in the apiserver.

@crassirostris @liggitt - Thoughts? Is this increase expected or maybe I should look for sth else?

**Comment by jpbetz** [comment]
@cheftako Any chance some of your recent memory analysis on apiserver might be related/helpful here?

**Comment by crassirostris** [comment]
This is an expected increase, audit logging now uses some memory to accumulate audit events before writing them to disk. No more than 10k events are stored by default

**Comment by shyamjvs** [comment]
Thanks @crassirostris.

> No more than 10k events are stored by default

So seems like that'd be a bounded increase in mem-usage (and independent of cluster size).
Could you confirm that and also how much bound wrt memory can we expect, so we can change our thresholds?

**Comment by shyamjvs** [comment]
Similarly, I fetched the usage for controller-manager too over the long-run:

![wider_cm_mem](https://user-images.githubusercontent.com/4333971/37306350-ba2618ec-2637-11e8-850a-b625c7ae9896.png)

And narrowed it down to:

![narrower_cm_range](https://user-images.githubusercontent.com/4333971/37306379-cc69e8a8-2637-11e8-8407-6d963dd2592d.png)

**Comment by shyamjvs** [comment]
Looking at the [diff](https://github.com/kubernetes/kubernetes/compare/d336607679743a...9a8b675d2c19fa) b/w runs 10868 and 10870 (which seems to be where the regression happened), I'm not really sure which one would cause this. Though this seems potential to me:

Try longer to fetch initial token. #56394

While it's making some POST requests internally and storing the response, it seems to be called as a one-off in the cloud-provider pkg (so not sure if it should increase mem-usage - maybe it's allocating some memory). I'll need to verify.

**Comment by shyamjvs** [comment]
I've changed the issue title to more accurately represent the situation. Turns out it's not as bad as 1.5-2x increase (as was originally noted).

**Comment by crassirostris** [comment]
@shyamjvs 

> Could you confirm that and also how much bound wrt memory can we expect, so we can change our thresholds?

Depends on the request sizes. Assuming each request in ~2.5KB (something I observed in e2e correctness tests), that's at least +25MB for buffering. Accounted for the different size internally, gc and the fact that a copy of audit logging request is stored in context for each request, +100-200 MB sounds reasonable.

**Comment by shyamjvs** [comment]
Thanks @crassirostris - I'll update our thresholds correspondingly. Could you please make a release-note mentioning this (if not already done) as this can affect end users?

**Comment by crassirostris** [comment]
@shyamjvs 

> Could you please make a release-note mentioning this (if not already done) as this can affect end users?

To clarify, advanced audit logging is enabled by default everywhere but in scale tests since 1.9 (or 1.8 even). That's not a new change per se. However, I created a PR to address this point in the documentation: https://github.com/kubernetes/website/pull/7725

**Comment by shyamjvs** [comment]
> To clarify, advanced audit logging is enabled by default everywhere but in scale tests since 1.9 (or 1.8 even).

I'm aware of that :) However, note that for our 100-node scalability jobs we [did not disable](https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gci-gce-scalability.env) advanced audit-logging (unlike for our [larger jobs](https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gce-large-performance.env#L26)). IIUC that's the reason why we caught this issue?

> That's not a new change per se.

While I agree that advanced audit-logging is not a new change, the buffering change (https://github.com/kubernetes/kubernetes/pull/60237) which caused this increase was introduced in 1.10, no?

**Comment by crassirostris** [comment]
> I'm aware of that :) However, note that for our 100-node scalability jobs we did not disable advanced audit-logging (unlike for our larger jobs). IIUC that's the reason why we caught this issue.
> While I agree that advanced audit-logging is not a new change, the buffering change (#60237) which caused this increase was introduced in 1.10, no?

Oh, OK. I got confused with all different scale tests, sorry :) I agree, PR in question did increase memory consumption on its own. Not sure it requires a separate release note or a piece of documentation, since the memory consumption is extremely dependent on the configuration. For example, buffering existed in webhook audit backend, so e.g. in GKE memory consumption shouldn't have changed.

> unlike for our larger jobs

BTW, I think this is a mistake, we can enable audit logging everywhere now, similarly to https://github.com/kubernetes/test-infra/pull/7000

**Comment by shyamjvs** [comment]
No problem :). Few thoughts:

> Not sure it requires a separate release note or a piece of documentation, since the memory consumption is extremely dependent on the configuration.

The documentation PR you created is definitely useful. However, my feeling is this also warrants a release-note. If it SGTY, I can add a release-note in the PR where I'll bump our test thresholds.

> since the memory consumption is extremely dependent on the configuration.

That's true. Maybe a rough idea of the upper bound (200MB?) with default config (i.e 10k events?) is enough.

> For example, buffering existed in webhook audit backend, so e.g. in GKE memory consumption shouldn't have changed.

For GKE, yup. One question though - did we change the k8s apiserver default to buffering mode in 1.10?

> BTW, I think this is a mistake, we can enable audit logging everywhere now, similarly to kubernetes/test-infra#7000

Yeah, seems like the *-performance jobs got left out. I'll send out a PR later enabling it.

**Comment by crassirostris** [comment]
> If it SGTY, I can add a release-note in the PR where I'll bump our test thresholds

SG, thanks!

> did we change the k8s apiserver default to buffering mode in 1.10?

For log backend, no. Or, rather yes, but there's a PR to revert it and re-introduce it in 1.11

> I'll send out a PR later enabling it.

Thanks!

**Comment by shyamjvs** [comment]
So the apiserver mem issue is sorted now - I've sent #61118 to bump the threshold.

And wrt the controller-manager mem issue, I've confirmed (by running density test few times against 100-node cluster) that it's being caused by this PR:

```
fix fluentd-gcp-scaler to look at correct fluentd-gcp version #59921
```

While running only density test,

- before that PR, the mem-usage is consistently ~300MB
- after that PR, the mem-usage is consistently ~330MB

@x13n @crassirostris - Any leads here? Maybe the fluentd-scaler wasn't doing any real scaling earlier due to wrong deployment name (just a guess though)?

**Comment by crassirostris** [comment]
@x13n should know more about the connection between scaler and controller-manager. I though all work is performed by a dedicated controller running in the cluster, not controller-manager

**Comment by x13n** [comment]
@shyamjvs Yes, before that PR it wasn't doing anything useful. Between https://github.com/kubernetes/kubernetes/pull/59657 and https://github.com/kubernetes/kubernetes/pull/59921, fluentd-gcp had no resource request nor limit set. Was there a 30MB drop in memory usage in https://github.com/kubernetes/kubernetes/pull/59657?

**Comment by shyamjvs** [comment]
Seems both those PRs were merged 27 days ago. However, we started seeing the increase only after #59921 (which IIUC is when the scaler actually started scaling sth?)
In general, do you have a reason to expect controller-manager to be affected by your addon?

**Comment by shyamjvs** [comment]
@x13n and me discussed about this offline and there are a bunch of possibilities why this might be happening (possibly some bad behaviour of the scaler).
I've shared instructions on how to run the 100-node tests with him and he'll dig deeper into this.

**Comment by shyamjvs** [comment]
@x13n Btw - you might find the following controller-manager logs:

- before your change: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/10865/artifacts/e2e-big-master/kube-controller-manager.log
- after your change: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/10870/artifacts/e2e-big-master/kube-controller-manager.log

Searching for 'fluentd', shows a bunch of such logs in the latter:

```
I0215 17:38:32.683798       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-qkbsq
I0215 17:38:32.684117       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-p4t52
I0215 17:38:32.684277       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-gs956
I0215 17:38:32.684444       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-xcqzw
I0215 17:38:32.684595       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-tmt9n
I0215 17:38:32.684736       1 controller_utils.go:590] Controller fluentd-gcp-v2.0.14 deleting pod kube-system/fluentd-gcp-v2.0.14-6js9l
```

which doesn't seem too normal to happen in b/w the test.

**Comment by x13n** [comment]
The first PR removed resource requests and limits from the manifest file used by addon manager. The second PR caused the scaler to set them again. Current hypothesis is that sometimes, when the scaler runs `kubectl set resources`, the resource is updated even though it shouldn't. This in turn causes controller-manager to receive data via watch on the DaemonSet resource, which causes the increase in memory usage.
I'm going to verify this and if this is the case, will add a check to the scaler so that set resources call is used only when necessary.

**Comment by x13n** [comment]
We did some debugging with @shyamjvs. What we found out is: addon manager and fluent-scaler are both sending PATCH requests to apiserver, to update fluentd-gcp. All these patches are no-ops, the only thing that is changing is the resource version. That unfortunately doesn't prevent the deamonset controller from killing & recreating identical pods, which is what causes increased memory usage of controller-manager.

Short term, I will change fluentd-scaler to send resource update requests only when there is actual change that needs to be PATCHed. However, empty patches shouldn't cause daemonset-controller to kill  any pods. The latter is probably not a release blocker (?), but should still be investigated.

**Comment by liggitt** [comment]
> All these patches are no-ops, the only thing that is changing is the resource version.

no-op patches should not change resource version or trigger watch events.

**Comment by shyamjvs** [comment]
@liggitt Even my feeling is the same that we shouldn't be changing RVs for no-ops. There's sth else in play here - needs more digging.

To summarize, here's what we observed:
1. PATCH daemonset calls coming every minute from both fluentd-scaler and addon-manager (verified this by individually turning them on/off). Things we need to understand here:
  - Is the daemonset object continuously toggling b/w 2 states? (we know here that it's RV is increasing continuously)
  - If yes, what field(s) in the object are changing? IIRC value of some label/annotation (I think 'UpdatedPodsScheduled') is changing (probably related to 2 below)
  - Also, why/should the fluentd-scaler send any api request if the resources are already set to the right value?
2. Fluentd pods are getting deleted and recreated by daemonset-controller when the scaler is enabled (as was also seen in https://github.com/kubernetes/kubernetes/issues/60500#issuecomment-373001797). Why this is happening? One thing to note here - all those delete calls are preceded by `PUT pod-status` calls from respective kubelets (but maybe that's expected).

**Comment by shyamjvs** [comment]
Summarizing the entire issue:
- Apiserver mem-usage increased: We found the reason and it's expected ([ref](https://github.com/kubernetes/kubernetes/issues/60500#issuecomment-372682659)). I sent a PR increasing the threshold for our scalability tests along with a release note explaining the effect (https://github.com/kubernetes/kubernetes/pull/61118) - it's now merged.
- Controller-manager mem-usage increased: We found the reason ([ref](https://github.com/kubernetes/kubernetes/issues/60500#issuecomment-372821309)) and it exposed a bug with fluentd-scaler. I've filed https://github.com/kubernetes/kubernetes/issues/61190 against sig-instrumentation to separately track this (as a release-blocker).
- Flakes we're generally seeing with our density test: I've filed https://github.com/kubernetes/kubernetes/issues/61189 to separately track it (currently a non-blocker).

IMO we can now close this issue and follow-up individually on those.

@jdumars @jberkus
